src/tools/idlc/idlc.c

```diff
@@ -199,56 +195,116 @@ static int idlc_printf(OUTDEST od, const char *fmt, ...)
     break;
   }
 
-  ddsrt_free(str);
+  free(str);
 
   return ret < 0 ? -1 : ret;
 }
 
-int32_t idlc_parse(ddsts_type_t **typeptr)
+static idl_retcode_t figure_file(idl_file_t **filep)
+{
+  idl_file_t *file;
+  char *dir = NULL, *abs = NULL, *name = NULL, *norm = NULL;
+  const char *sep = NULL, *ptr;
+
+  if (!(file = malloc(sizeof(*file))))
+    goto err_file;
+  /* determine "relative" file name */
+  for (ptr=opts.file; *ptr; ptr++) {
+    if (idl_isseparator(*ptr))
+      sep = ptr;
+  }
+  if (!(name = idl_strdup(sep ? sep : opts.file)))
+    goto err_name;
+  if (idl_isabsolute(opts.file)) {
+    if (idl_normalize_path(opts.file, &norm) < 0)
+      goto err_file;
+  } else {
+    if (idl_current_path(&dir) < 0)
+      goto err_dir;
+    if (idl_asprintf(&abs, "%s/%s", dir, opts.file) == -1)
+      goto err_abs;
+    if (idl_normalize_path(abs, &norm) < 0)
+      goto err_norm;
+    free(abs);
+  }
+  file->next = NULL;
+  file->name = name;
+  *filep = file;
+  return IDL_RETCODE_OK;
+err_norm:
+  free(abs);
+err_abs:
+  free(dir);
+err_dir:
+  free(name);
+err_name:
+  free(file);
+err_file:
+  return IDL_RETCODE_NO_MEMORY;
+}
+
+static int32_t idlc_parse(void) //idl_tree_t **treeptr)
 {
+  idl_file_t *file = NULL;
+  //idl_tree_t *tree;
   int32_t ret = 0;
 
+  //if (!(tree = calloc(1, sizeof(*tree))))
+  //  return IDL_RETCODE_NO_MEMORY;
+
   if(opts.flags & IDLC_COMPILE) {
-    if ((ret = idl_processor_init(&proc)) != 0)
+    if ((ret = idl_create_pstate(0u, NULL, &pstate))) { //rocessor_init(&proc)) != 0) {
+      //free(tree);
       return ret;
+    }
     assert(opts.file);
-    if (strcmp(opts.file, "-") != 0)
-      proc.scanner.position.file = (const char *)opts.file;
-    proc.scanner.position.line = 1;
-    proc.scanner.position.column = 1;
+    if (strcmp(opts.file, "-") != 0 && (ret = figure_file(&file)) != 0) {
+      idl_delete_pstate(pstate);
+      //idl_processor_fini(&proc);
+      //free(tree);
+      return ret;
+    }
+    pstate->files = file;
+    pstate->scanner.position.file = (const idl_file_t *)file;
+    pstate->scanner.position.line = 1;
+    pstate->scanner.position.column = 1;
+    pstate->flags |= IDL_WRITE;
   }
 
   if (opts.flags & IDLC_PREPROCESS) {
-    proc.flags |= IDL_WRITE;
+    pstate->flags |= IDL_WRITE | IDL_FLAG_ANNOTATIONS | IDL_FLAG_EXTENDED_DATA_TYPES;
```

Comment:
`pstate` is not initialized at this point when only running preprocessor

Reply:
fixed